packages/core/src/util/esm.js

```diff
@@ -39,3 +39,34 @@ exports.hasExperimentalLoaderFlag = function hasExperimentalLoaderFlag() {
 
   return experimentalLoaderFlagIsSet && exports.isLatestEsmSupportedVersion(process.versions.node);
 };
+
+/**
+ * Checks if the application is an ECMAScript Modules (ESM) app by inspecting
+ * the presence of flags like --experimental-loader and --import in the environment.
+ * @returns {boolean} True if an ESM app is detected, false otherwise.
+ * If the application is ESM, caching the result for efficiency.
+ */
+exports.isESMApp = function isESMApp() {
+  const cacheKey = 'isESMApp';
+
+  if (require.cache[cacheKey]) {
+    return require.cache[cacheKey].exports;
+  }
+  const isESM =
+    (process.env.NODE_OPTIONS &&
+      (process.env.NODE_OPTIONS.indexOf('--experimental-loader') !== -1 ||
+        process.env.NODE_OPTIONS.indexOf('--import') !== -1)) ||
+    (process.execArgv &&
+      process.execArgv.length > 0 &&
+      ((process.execArgv[0].indexOf('--experimental-loader') !== -1 &&
+        process.execArgv[0].indexOf('esm-loader.mjs') !== -1) ||
+        (process.execArgv[0].indexOf('--experimental-loader') !== -1 &&
+          process.execArgv[1].indexOf('esm-loader.mjs') !== -1) ||
+        (process.execArgv[0].indexOf('--import') !== -1 && process.execArgv[0].indexOf('esm-register.mjs') !== -1) ||
+        (process.execArgv[0].indexOf('--import') !== -1 && process.execArgv[1].indexOf('esm-register.mjs') !== -1)));
+
+  // @ts-ignore
+  require.cache[cacheKey] = { exports: isESM };
```

Comment:
???

Reply:
require.cache is a file cache. Please do not abuse it for caching a local variable.